### ISSUE TITLE
Pyright sending client/registerCapability even if dynamicRegistration set to false during initialization

### ISSUE THREAD

**Issue by mjlbach** [microsoft/pyright#1302]
**Describe the bug**

We have support for pyright as one of the language servers for neovim's built in language client. One issue seems to be cropping up on node 15 where failed callbacks result in node terminating the process. The main callback that is failing is client/registerCapability, (our client doesn't support dynamic registration). When we report dynamicRegistration as false, pyright is still sending client/registerCapability requests

```
[ DEBUG ] 2020-12-20T13:55:08-0800 ] /home/michael/neovim/share/nvim/runtime/lua/vim/lsp/rpc.lua:489 ]	
"decoded"	{  id = 0,  jsonrpc = "2.0",  method = "client/registerCapability",  params = {    registrations = { {        
id = "807eedda-04cf-4637-9cb7-81d5f5d0e225",        method = "workspace/didChangeWorkspaceFolders",       
 registerOptions = vim.empty_dict()      } }  }}
```
We're initializing the server with all dynamicRegistration capabilities registered as false

<details open>
<summary>json of initialization params</summary>

```json
{
  "initialize_params": {
    "capabilities": {
      "callHierarchy": {
        "dynamicRegistration": false,
        "textDocument": {
          "codeAction": {
            "codeActionLiteralSupport": {
              "codeActionKind": {
                "valueSet": [
                  "",
                  "Empty",
                  "QuickFix",
                  "Refactor",
                  "RefactorExtract",
                  "RefactorInline",
                  "RefactorRewrite",
                  "Source",
                  "SourceOrganizeImports",
                  "quickfix",
                  "refactor",
                  "refactor.extract",
                  "refactor.inline",
                  "refactor.rewrite",
                  "source",
                  "source.organizeImports"
                ]
              }
            },
            "dynamicRegistration": false
          },
          "completion": {
            "completionItem": {
              "commitCharactersSupport": false,
              "deprecatedSupport": false,
              "documentationFormat": [
                "markdown",
                "plaintext"
              ],
              "preselectSupport": false,
              "snippetSupport": false
            },
            "completionItemKind": {
              "valueSet": [
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
                10,
                11,
                12,
                13,
                14,
                15,
                16,
                17,
                18,
                19,
                20,
                21,
                22,
                23,
                24,
                25
              ],
              "contextSupport": false,
              "dynamicRegistration": false
            },
            "declaration": {
              "linkSupport": true
            },
            "definition": {
              "linkSupport": true
            },
            "documentHighlight": {
              "dynamicRegistration": false
            },
            "documentSymbol": {
              "dynamicRegistration": false,
              "hierarchicalDocumentSymbolSupport": true,
              "symbolKind": {
                "valueSet": [
                  1,
                  2,
                  3,
                  4,
                  5,
                  6,
                  7,
                  8,
                  9,
                  10,
                  11,
                  12,
                  13,
                  14,
                  15,
                  16,
                  17,
                  18,
                  19,
                  20,
                  21,
                  22,
                  23,
                  24,
                  25,
                  26
                ]
              }
            },
            "hover": {
              "contentFormat": [
                "markdown",
                "plaintext"
              ],
              "dynamicRegistration": false
            },
            "      implementation": {
              "        linkSupport": true
            },
            "references": {
              "dynamicRegistration": false
            },
            "rename": {
              "dynamicRegistration": false,
              "prepareSupport": true
            },
            "signatureHelp": {
              "dynamicRegistration": false,
              "signatureInformation": {
                "documentationFormat": [
                  "markdown",
                  "plaintext"
                ]
              }
            },
            "      synchronization": {
              "didSave": true,
              "dynamicRegistration": false,
              "willSave": false,
              "willSaveWaitUntil": false
            },
            "      typeDefinition": {
              "        linkSupport": true
            }
          },
          "workspace": {
            "applyEdit": true,
            "configuration": true,
            "symbol": {
              "dynamicRegistration": false,
              "hierarchicalWorkspaceSymbolSupport": true,
              "symbolKind": {
                "valueSet": [
                  1,
                  2,
                  3,
                  4,
                  5,
                  6,
                  7,
                  8,
                  9,
                  10,
                  11,
                  12,
                  13,
                  14,
                  15,
                  16,
                  17,
                  18,
                  19,
                  20,
                  21,
                  22,
                  23,
                  24,
                  25,
                  26
                ]
              }
            },
            "workspaceFolders": true
          }
        },
        "initializationOptions": {},
        "processId": 53558,
        "rootPath": "/home/michael/Repositories/pytorch_learning",
        "rootUri": "file:///home/michael/Repositories/pytorch_learning",
        "trace": "off",
        "workspaceFolders": [
          {
            "name": "/home/michael/Repositories/pytorch_learning",
            "uri": "file:///home/michael/Repositories/pytorch_learning"
          }
        ]
      }
    }
  }
}
```
</details>

**To Reproduce**
This can't be reproduced in vscode. I believe lsp-mode for emacs also implements dynamic registration, so the only way to reproduce this is to install neovim and nvim-lspconfig on node 14 (which doesn't crash). I know this is out of scope for pyright so I will do all of the legwork on fixing this.

**Expected behavior**
Pyright doesn't send registerCapability requests to the client if dynamicRegistration is set to false (or unset)

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

**Additional context**
As mentioned I'm happy to submit a PR to fix this if acceptable, I've just had trouble finding exactly where in the pyright languageserver extension dynamicRegistration is read from initialization params, and where the registerCapability request is triggered.

**Comment by erictraut** [comment]
As far as I know, the Pyright code performs a dynamic registration in only one place. It's on [this line](https://github.com/microsoft/pyright/blob/64408a5cf0fdd0ba8f8f78baca785962f8e507d3/packages/pyright-internal/src/languageServerBase.ts#L841). This registration call is gated by a check for `_hasFileWatchCapability`, which is initialized as follows:

```typescript
        this._hasWatchFileCapability = !!capabilities.workspace?.didChangeWatchedFiles?.dynamicRegistration;
```

So it appears we're checking for the `dynamicRegistration` flag from the client in this case. Can you confirm that your client is sending false for this expression?

**Comment by mjlbach** [comment]
Yes, I just confirmed it. Sorry for the poor formatting (I did the above by hand)

```
[ DEBUG ] 2020-12-20T15:56:46-0800 ] /home/michael/neovim/share/nvim/runtime/lua/vim/lsp/rpc.lua:388 ]	"rpc.send.payload"	{  id = 1,  jsonrpc = "2.0",  method = "initialize",  params = {    capabilities = {      textDocument = {        callHierarchy = {          dynamicRegistration = false,          <metatable> = <1>{            __tostring = <function 1>          }        },        codeAction = {          codeActionLiteralSupport = {            codeActionKind = {              valueSet = { "", "Empty", "QuickFix", "Refactor", "RefactorExtract", "RefactorInline", "RefactorRewrite", "Source", "SourceOrganizeImports", "quickfix", "refactor", "refactor.extract", "refactor.inline", "refactor.rewrite", "source", "source.organizeImports" },              <metatable> = <table 1>            },            <metatable> = <table 1>          },          dynamicRegistration = false,          <metatable> = <table 1>        },        codeLens = {          dynamicRegistration = false,          <metatable> = <table 1>        },        colorProvider = {          dynamicRegistration = false,          <metatable> = <table 1>        },        completion = {          completionItem = {            commitCharactersSupport = false,            deprecatedSupport = false,            documentationFormat = { "markdown", "plaintext" },            preselectSupport = false,            snippetSupport = false,            <metatable> = <table 1>          },          completionItemKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25 },            <metatable> = <table 1>          },          contextSupport = false,          dynamicRegistration = false,          <metatable> = <table 1>        },        declaration = {          dynamicRegistration = false,          linkSupport = true,          <metatable> = <table 1>        },        definition = {          dynamicRegistration = false,          linkSupport = true,          <metatable> = <table 1>        },        documentHighlight = {          dynamicRegistration = false,          <metatable> = <table 1>        },        documentLink = {          dynamicRegistration = false,          <metatable> = <table 1>        },        documentSymbol = {          dynamicRegistration = false,          hierarchicalDocumentSymbolSupport = true,          symbolKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 },            <metatable> = <table 1>          },          <metatable> = <table 1>        },        foldingRange = {          dynamicRegistration = false,          <metatable> = <table 1>        },        formatting = {          dynamicRegistration = false,          <metatable> = <table 1>        },        hover = {          contentFormat = { "markdown", "plaintext" },          dynamicRegistration = false,          <metatable> = <table 1>        },        implementation = {          dynamicRegistration = false,          linkSupport = true,          <metatable> = <table 1>        },        linkedEditingRange = {          dynamicRegistration = false,          <metatable> = <table 1>        },        onTypeFormatting = {          dynamicRegistration = false,          <metatable> = <table 1>        },        rangeFormatting = {          dynamicRegistration = false,          <metatable> = <table 1>        },        references = {          dynamicRegistration = false,          <metatable> = <table 1>        },        rename = {          dynamicRegistration = false,          prepareSupport = true,          <metatable> = <table 1>        },        selectionRange = {          dynamicRegistration = false,          <metatable> = <table 1>        },        semanticTokens = {          dynamicRegistration = false,          <metatable> = <table 1>        },        signatureHelp = {          dynamicRegistration = false,          signatureInformation = {            documentationFormat = { "markdown", "plaintext" },            <metatable> = <table 1>          },          <metatable> = <table 1>        },        synchronization = {          didSave = true,          dynamicRegistration = false,          willSave = false,          willSaveWaitUntil = false,          <metatable> = <table 1>        },        textDocumentSync = {          dynamicRegistration = false,          <metatable> = <table 1>        },        typeDefinition = {          dynamicRegistration = false,          linkSupport = true,          <metatable> = <table 1>        },        <metatable> = <table 1>      },      workspace = {        applyEdit = true,        codeLens = {          dynamicRegistration = false,          <metatable> = <table 1>        },        configuration = false,        didChangeConfiguration = {          dynamicRegistration = false,          <metatable> = <table 1>        },        didChangeWatchedFiles = {          dynamicRegistration = false,          <metatable> = <table 1>        },        executeCommand = {          dynamicRegistration = false,          <metatable> = <table 1>        },        fileOperations = {          dynamicRegistration = false,          <metatable> = <table 1>        },        semanticTokens = {          dynamicRegistration = false,          <metatable> = <table 1>        },        symbol = {          dynamicRegistration = false,          hierarchicalWorkspaceSymbolSupport = true,          symbolKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 },            <metatable> = <table 1>          },          <metatable> = <table 1>        },        workspaceFolders = true,        <metatable> = <table 1>      }    },    initializationOptions = vim.empty_dict(),    processId = 4074,    rootPath = "/home/michael/Repositories/pytorch_learning",    rootUri = "file:///home/michael/Repositories/pytorch_learning",    trace = "off",    workspaceFolders = { {        name = "/home/michael/Repositories/pytorch_learning",        uri = "file:///home/michael/Repositories/pytorch_learning"      } }  }}
[ DEBUG ] 2020-12-20T15:56:46-0800 ] /home/michael/neovim/share/nvim/runtime/lua/vim/lsp/rpc.lua:489 ]	"decoded"	{  jsonrpc = "2.0",  method = "window/logMessage",  params = {    message = "Pyright language server 1.1.95 starting",    type = 3  }}
[ DEBUG ] 2020-12-20T15:56:46-0800 ] /home/michael/neovim/share/nvim/runtime/lua/vim/lsp/rpc.lua:388 ]	"rpc.send.payload"	{  jsonrpc = "2.0",  method = "textDocument/didOpen",  params = {    textDocument = {      languageId = "python",      text = 'import collections\nimport numpy as np\n\ndef main():\n    print("Test")\n\ndef test():\n    print("True")\n',      uri = "file:///home/michael/Repositories/pytorch_learning/main.py",      version = 0    }  }}
[ DEBUG ] 2020-12-20T15:56:46-0800 ] /home/michael/neovim/share/nvim/runtime/lua/vim/lsp/rpc.lua:489 ]	"decoded"	{  id = 0,  jsonrpc = "2.0",  method = "client/registerCapability",  params = {    registrations = { {        id = "b2d5bb77-4a26-4d8f-ac67-0f8bef623ef9",        method = "workspace/didChangeWorkspaceFolders",        registerOptions = vim.empty_dict()      } }  }}
```

The response seems to indicate workspace?.didChangeWatchedFiles?.dynamicRegistration is set to false, but I still get client/registerCapability being triggered

**Comment by erictraut** [comment]
The code you pasted above is in the VS Code extension, not in the LSP, so I wouldn't expect that code to be involved.

**Comment by erictraut** [comment]
The configuration you pasted above looks fine to me, so I don't understand how we could be going down the code path where the dynamic registration is created. You're in a better position to debug this than I am, so I welcome your help in finding the problem.

**Comment by mjlbach** [comment]
Oh, my mistake. I just noticed that as well after building the modified version. I'm going to look into this and I appreciate the debugging advice!

**Comment by mjlbach** [comment]
I'm going to close this as I'm fairly certain at this point the problem lies in vscode-languageserver-node rather than in pyright itself.
